### PR TITLE
Use same code to complete GAV at different levels

### PR DESCRIPTION
**Pull request by mickaelistria**
This also brings
* parent GAV completion when relative path is set
* completing inside `dependency` and `plugin`

Signed-off-by: Mickael Istria <mistria@redhat.com>

### CODE REVIEW

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenCompletionParticipant.java [hunk 1]
```diff
@@ -212,6 +261,40 @@ public void onXMLContent(ICompletionRequest request, ICompletionResponse respons
...
+		Optional<String> relativePath = null;
+		if (request.getParentElement().getLocalName().equals("parent")) {
+			relativePath = DOMUtils.findChildElementText(request.getNode(), "relativePath");
+		} else {
+			relativePath = DOMUtils.findChildElementText(request.getParentElement().getParentElement(), "relativePath");
+		}
+		if (!relativePath.isPresent()) {
+			relativePath = Optional.of("..");
+		}
+		File referencedTargetPomFile = new File(new File(URI.create(request.getXMLDocument().getTextDocument().getUri())).getParentFile(), relativePath.orElse(""));
```
Comment (AObuchow) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenCompletionParticipant.java]: Just to make sure I follow:
1. We check if the <relativePath> was writen by the user
2. We then create a grandparent file of the current XML document (which is most likely a directory) or create a file using the relative path
3. If the file created in step 2 was a directory, create a File object for a pom.xml in that directory (or if it's already a file then move to step 4)
4. Parse the file object and return it

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenCompletionParticipant.java [hunk 2]
```diff
@@ -212,6 +261,40 @@ public void onXMLContent(ICompletionRequest request, ICompletionResponse respons
 		}
 	}
 
+	private InsertionStrategy computeGAVInsertionStrategy(ICompletionRequest request) {
```
Comment (AObuchow) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenCompletionParticipant.java]: This is so nice, I really like this abstraction.

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenProjectCache.java [hunk 3]
```diff
@@ -107,12 +105,41 @@ public MavenProject getLastSuccessfulMavenProject(DOMDocument document) {
...
 		Integer last = lastCheckedVersion.get(URI.create(document.getTextDocument().getUri()));
 		if (last == null || last.intValue() < document.getTextDocument().getVersion()) {
-			parse(document);
+			parseAndCache(document);
 
 		}
 	}
 
-	private void parse(DOMDocument document) {
+	public Optional<MavenProject> getSnspahotProject(File file) {
```
Comment (AObuchow) [inline lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenProjectCache.java]: Small typo, should be `getSnapshotProject(File file)`

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenCompletionParticipant.java (hunk 1 above)
Comment (mickaelistria) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenCompletionParticipant.java]: If relativePath is not set, we assume it's `..` (the parent folder).

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenCompletionParticipant.java (hunk 2 above)
Comment (mickaelistria) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenCompletionParticipant.java]: It's inspired from the "strategy" design pattern, but it's not as good as the actual pattern ;)

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenProjectCache.java (hunk 3 above)
Comment (mickaelistria) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenProjectCache.java]: Thanks

Comment (mickaelistria) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenProjectCache.java]: Fixed in new commit

File: lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenCompletionParticipant.java (hunk 2 above)
Comment (AObuchow) [reply lemminx-maven/src/main/java/org/eclipse/lemminx/maven/MavenCompletionParticipant.java]: I like it a lot regardless :)

**Comment by AObuchow** [comment]
Awesome patch, thanks @mickaelistria !
Good to merge.

**Pull request merged**